daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/ChoiceCombinator.scala

```diff
@@ -83,6 +84,15 @@ case class ChoiceCombinator(ch: ChoiceTermBase, alternatives: Seq[Gram])
         case ChoiceLengthKind.Explicit => new SpecifiedLengthChoiceParser(cp, ch.choiceRuntimeData, choiceLengthInBits.get)
       }
     } else {
+      //Verify that every alternative has some form of branch key
+      alternatives.map{ alt =>
+        val keyTerm = alt.context.asInstanceOf[Term]
+        val hasBranchKey = keyTerm.findPropertyOption("choiceBranchKey").isDefined
+        val hasBranchKeyRanges = keyTerm.findPropertyOption("choiceBranchKeyRanges").isDefined
+        if(!hasBranchKey && !hasBranchKeyRanges && ch.defaultableChoiceBranchKeyKind != ChoiceKeyKindType.ByType){
+          keyTerm.SDE("dfdl:choiceBranchKey and dfdlx:choiceBranchKeyRanges are not defined.")
```

Comment:
Should say "Neither dfdl:choiceBranchKey nor dfdlx:choiceBranchKeyRanges is defined." since these are alternatives and neither-nor is singular. Of course I realize they aren't exclusive alternatives. One can have both, but neither-nor is still less ambiguous than "and".